### PR TITLE
bpo-44011: New asyncio ssl implementation

### PR DESCRIPTION
**Pull request by asvetlov**
Borrowed from uvloop

https://bugs.python.org/issue44011

### CODE REVIEW

**Comment by asvetlov** [comment]
The code part is finished; seems work ok.
I want to find time for porting a part of functional tests from uvloop.

**Comment by 1st1** [comment]
>  I want to find time for porting a part of functional tests from uvloop.

That's the key and should be relatively easy, since the tests already test asyncio in many cases.

File: Lib/asyncio/sslproto.py [hunk 1]
```diff
@@ -238,11 +238,8 @@ def abort(self):
         The protocol's connection_lost() method will (eventually) be
         called with None as its argument.
         """
-        self._force_close(None)
-
-    def _force_close(self, exc):
```
Comment (fantix) [inline Lib/asyncio/sslproto.py]: I think `_force_close` made it possible to do SSL over SSL, so we should probably keep it.

**Review by fantix — COMMENTED** [review]
Thanks a lot for working on this! The rest looks good to me.

File: Lib/asyncio/sslproto.py (hunk 1 above)
Comment (asvetlov) [reply Lib/asyncio/sslproto.py]: Good catch!
Done

**Comment by 1st1** [comment]
Oh, looks like we failed to include this in 3.9. @fantix can we commit it at least now?

**Comment by 1st1** [comment]
Also pinging @asvetlov

**Comment by fantix** [comment]
Yes, I believe it's fine to move on. The only missing part is

> porting a part of functional tests from uvloop

I'll try to help with the porting when time.

**Comment by dimaqq** [comment]
@asvetlov would you consider rebasing your branch?
I can't compile it in the current state on macOS with:

```
./Modules/posixmodule.c:9302:15: error: implicit declaration of function 'sendfile' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        ret = sendfile(in, out, offset, &sbytes, &sf, flags);
```

Which is not among your changes.

**Comment by fantix** [comment]
@dimaqq I'll give it a try this weekend - there're new changes to be added here: https://github.com/MagicStack/uvloop/pull/385

**Comment by tiran** [comment]
This is a very large PR. I don't think I'll be able give you a proper review until feature freeze.

We should get together and figure out which features of OpenSSL I can expose to improve the asyncio implementation. For example OpenSSL 1.1.1 has a proper state machine.

**Comment by asvetlov** [comment]
@tiran the PR is a backport of uvloop battle-tested implementation.

@fantix (the author of this part of uvloop) already reviewed it.
We missed the time window for Python 3.9. I've added the required tests and hope to land it before 3.10 beta.

I'm open for new OpenSSL-related improvements but the PR makes a value as is. The current SSL/TLS implementation in asyncio is subtle and error-prone, trust me. Ad this PR is a way to fix the current state.

**Comment by tiran** [comment]
Go for it! :)

I now realized that I got auto-assigned as reviewer because one of the files contains the string "ssl".

**Comment by asvetlov** [comment]
Huh, constantly forgetting about `#` -> `GH-` renaming.

**Pull request merged**